File: system-test/java/src/main/java/com/optivem/shop/testkit/driver/adapter/shop/api/ShopApiDriver.java
package com.optivem.shop.testkit.driver.adapter.shop.api;

import com.optivem.shop.testkit.driver.adapter.shop.api.client.ShopApiClient;
import com.optivem.shop.testkit.driver.port.shop.dtos.error.SystemError;
import com.optivem.shop.testkit.driver.port.shop.dtos.BrowseCouponsResponse;
import com.optivem.shop.testkit.driver.port.shop.dtos.PlaceOrderRequest;
import com.optivem.shop.testkit.driver.port.shop.dtos.PlaceOrderResponse;
import com.optivem.shop.testkit.driver.port.shop.dtos.PublishCouponRequest;
import com.optivem.shop.testkit.driver.port.shop.dtos.ViewOrderResponse;
import com.optivem.shop.testkit.driver.port.shop.ShopDriver;
import com.optivem.shop.testkit.common.Closer;
import com.optivem.shop.testkit.common.Result;

public class ShopApiDriver implements ShopDriver {
    private final ShopApiClient apiClient;

    public ShopApiDriver(String baseUrl) {
        this.apiClient = new ShopApiClient(baseUrl);
    }

    @Override
    public void close() {
        Closer.close(apiClient);
    }

    @Override
    public Result<Void, SystemError> goToShop() {
        return apiClient.health().checkHealth().mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<PlaceOrderResponse, SystemError> placeOrder(PlaceOrderRequest request) {
        return apiClient.orders().placeOrder(request).mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<Void, SystemError> cancelOrder(String orderNumber) {
        return apiClient.orders().cancelOrder(orderNumber).mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<Void, SystemError> deliverOrder(String orderNumber) {
        return apiClient.orders().deliverOrder(orderNumber).mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<ViewOrderResponse, SystemError> viewOrder(String orderNumber) {
        return apiClient.orders().viewOrder(orderNumber).mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<Void, SystemError> publishCoupon(PublishCouponRequest request) {
        return apiClient.coupons().publishCoupon(request).mapError(SystemErrorMapper::from);
    }

    @Override
    public Result<BrowseCouponsResponse, SystemError> browseCoupons() {
        return apiClient.coupons().browseCoupons().mapError(SystemErrorMapper::from);
    }
}

File: system-test/typescript/src/testkit/driver/adapter/shop/api/shop-api-driver.ts
import type { Result } from '../../../../common/result.js';
import type { PlaceOrderRequest } from '../../../port/shop/dtos/PlaceOrderRequest.js';
import type { PlaceOrderResponse } from '../../../port/shop/dtos/PlaceOrderResponse.js';
import type { ViewOrderResponse } from '../../../port/shop/dtos/ViewOrderResponse.js';
import type { SystemError } from '../../../port/shop/dtos/errors/SystemError.js';
import type { PublishCouponRequest } from '../../../port/shop/dtos/PublishCouponRequest.js';
import type { BrowseCouponsResponse } from '../../../port/shop/dtos/BrowseCouponsResponse.js';
import type { ShopDriver } from '../../../port/shop/shop-driver.js';
import { ShopApiClient } from './client/ShopApiClient.js';

export class ShopApiDriver implements ShopDriver {
  private readonly client: ShopApiClient;

  constructor(baseUrl: string) {
    this.client = new ShopApiClient(baseUrl);
  }

  async goToShop(): Promise<Result<void, SystemError>> {
    return this.client.health().checkHealth();
  }

  async placeOrder(request: PlaceOrderRequest): Promise<Result<PlaceOrderResponse, SystemError>> {
    return this.client.orders().placeOrder(request);
  }

  async cancelOrder(orderNumber: string): Promise<Result<void, SystemError>> {
    return this.client.orders().cancelOrder(orderNumber);
  }

  async deliverOrder(orderNumber: string): Promise<Result<void, SystemError>> {
    return this.client.orders().deliverOrder(orderNumber);
  }

  async viewOrder(orderNumber: string): Promise<Result<ViewOrderResponse, SystemError>> {
    return this.client.orders().viewOrder(orderNumber);
  }

  async publishCoupon(request: PublishCouponRequest): Promise<Result<void, SystemError>> {
    return this.client.coupons().publishCoupon(request);
  }

  async browseCoupons(): Promise<Result<BrowseCouponsResponse, SystemError>> {
    return this.client.coupons().browseCoupons();
  }

  async close(): Promise<void> {}
}
